### PR TITLE
Removed lazy calculation and storage of norm

### LINKED ISSUE #25254: Quaternions in version 12
Dear friends,
I have detected a problem when multiplying Quaternions using the * to do the product between two.  It also fails when using q.mul(p).  In my case, one of the quaternions was obtained by the "from Axis angle". This code gives me problems:

`p= Quaternion(1,0,1,0);
q = Quaternion.from_axis_angle((1,1,1), 3*np.pi/4);
qi =q.inverse();
p1= q*p*qi `

The error is not there for any previous versions 11.x.

Many thanks for considering my request.

### PR DESCRIPTION
**Pull request by evbernardes**
#### References to other Issues or PRs
Fixes  #25254

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by sylee957** [comment]
Now, how does `self._norm` used for?
I don't think that related design choices like `set_norm` are meaningful for now, and could be deprecated, removed.

**Comment by evbernardes** [comment]
There are two cases: 
- you give a predetermined norm value as input when creating the quaternion. Then when asking for the norm, you just get this pre determined norm.
- you don't give a predetermined value and it calculates the norm.

Before this change, when calculating the norm at least once, the calculated norm would be stored as if it were a predefined norm and it would never be calculated again. The problem is that in this case, Quaternions that were not suppose to have a predetermined norm would be treated as such, creating potential problems for users who don't know about the norm system.

I just removed this in this case, but the norm system still exists for anyone interested in using it explicitly.

**Comment by sylee957** [comment]
You can use cacheit hook to cache the norm.
I'm just confused about the semantics that why users are able to 'set' the norm.

**Comment by sylee957** [comment]
I think that you need to add a test case for the steps of the original issue at least, because it's not obvious that the changes leads to the fix.

File: sympy/algebras/tests/test_quaternion.py [hunk 1]
```diff
@@ -46,6 +47,18 @@ def test_quaternion_construction_norm():
     assert (q1 * q3).norm() == q1.norm()
 
 
+def test_issue_25254():
+    # calculating the inverse cached the norm which caused problems
+    # when multiplying
+    p = Quaternion(1, 0, 1, 0)
+    q = Quaternion.from_axis_angle((1,1,1), 3 * np.pi/4)
```
Comment (sylee957) [inline sympy/algebras/tests/test_quaternion.py:54]: We can test with `math.pi` constant to remove dependency with numpy.

File: sympy/algebras/tests/test_quaternion.py [hunk 2]
```diff
@@ -46,6 +47,18 @@ def test_quaternion_construction_norm():
...
 
 
+def test_issue_25254():
+    # calculating the inverse cached the norm which caused problems
+    # when multiplying
+    p = Quaternion(1, 0, 1, 0)
+    q = Quaternion.from_axis_angle((1,1,1), 3 * math.pi/4)
+    qi = q.inverse()  # this operation cached the norm
+    try:
+        test = q * p * qi
```
Comment (sylee957) [inline sympy/algebras/tests/test_quaternion.py:57]: You can't assign the unused variable here.

Comment (evbernardes) [reply sympy/algebras/tests/test_quaternion.py]: I simplified the test with something equivalent that would also fail without this change for the same reasons.

Comment (evbernardes) [reply sympy/algebras/tests/test_quaternion.py]: Almost everything seems to be passing now, except for the tests related to Python 3.12. Is this enough?

**Pull request merged**